src/coprocessor/dag/executor/aggregation.rs

```diff
@@ -208,6 +209,181 @@ impl Executor for HashAggExecutor {
+    is_first_group: bool,
+    count: i64,
+}
+
+impl StreamAggExecutor {
+    pub fn new(
+        ctx: Arc<EvalContext>,
+        src: Box<Executor>,
+        mut meta: Aggregation,
+        columns: Arc<Vec<ColumnInfo>>,
+    ) -> Result<StreamAggExecutor> {
+        let mut visitor = ExprColumnRefVisitor::new(columns.len());
+        let group_bys = meta.take_group_by().into_vec();
+        visitor.batch_visit(&group_bys)?;
+        let aggs = meta.take_agg_func().into_vec();
+        visitor.batch_visit(&aggs)?;
+        let group_len = group_bys.len();
+        let exprs = AggFuncExpr::batch_build(&ctx, aggs)?;
+        // get aggregation functions
+        let mut funcs = Vec::with_capacity(exprs.len());
+        for expr in &exprs {
+            let agg = aggregate::build_aggr_func(expr.tp)?;
+            funcs.push(agg);
+        }
+
+        Ok(StreamAggExecutor {
+            src: src,
+            executed: false,
+            agg_exprs: exprs,
+            agg_funcs: funcs,
+            group_by_exprs: box_try!(Expression::batch_build(&ctx, group_bys)),
+            ctx: ctx,
+            related_cols_offset: visitor.column_offsets(),
+            cols: columns,
+            curr_group_row: vec![Datum::Null; group_len],
+            next_group_row: vec![Datum::Null; group_len],
+            tmp_group_row: vec![Datum::Null; group_len],
+            is_first_group: true,
+            count: 0,
+        })
+    }
+
+    // row: &[Datum]
+    fn meet_new_group(&mut self, row: &[Datum]) -> Result<bool> {
+        if self.group_by_exprs.is_empty() {
+            return Ok(false);
+        }
+
+        let mut matched = !self.is_first_group;
+        for (cnt, expr) in self.group_by_exprs.iter().enumerate() {
+            let v = box_try!(expr.eval(&self.ctx, row));
+            if matched && box_try!(v.cmp(&self.ctx, &self.next_group_row[cnt])) != Ordering::Equal {
+                matched = false;
+            }
+            self.tmp_group_row[cnt] = v;
+        }
+        let ret = !self.is_first_group;
+        if self.is_first_group {
+            self.curr_group_row = self.tmp_group_row.clone();
+            self.is_first_group = false
+        }
+        if matched {
+            return Ok(false);
+        }
+        self.next_group_row = self.tmp_group_row.clone();
+        Ok(ret)
+    }
+
+    fn get_partial_result(&mut self) -> Result<Option<Row>> {
+        let mut agg_cols = Vec::with_capacity(2 * self.agg_funcs.len());
+        // calc all agg func
+        for (cnt, agg_func) in self.agg_funcs.iter_mut().enumerate() {
+            agg_func.calc(&mut agg_cols)?;
+            let agg = aggregate::build_aggr_func(self.agg_exprs[cnt].tp)?;
+            *agg_func = agg;
+        }
+
+        // group by
+        let mut group_key = Vec::with_capacity(0);
+        if !self.group_by_exprs.is_empty() {
+            let mut vals = Vec::with_capacity(self.group_by_exprs.len());
+            for d in &self.curr_group_row {
```

Comment:
It seems we can consume `self.curr_group_row` here since we redefined it on L368?